coffee/utils.py

```diff
@@ -624,6 +592,184 @@ def to_for(self, itspaces, dims=None, stmts=None):
         return loops
 
 
+###############################################################
+# Utilities for tracking the global impact of transformations #
+###############################################################
+
+
+class StmtTracker(OrderedDict):
+
+    """Track the location of generic statements in an abstract syntax tree.
+
+    Each key in the dictionary is a string representing a symbol. As such,
+    StmtTracker can be used only in SSA scopes. Each entry in the dictionary
+    is a 4-tuple containing information about the symbol: ::
+
+        (statement, declaration, closest_for, place)
+
+    whose semantics is, respectively, as follows:
+
+        * The AST node whose ``str(lvalue)`` is used as dictionary key
+        * The AST node of the symbol declaration
+        * The AST node of the closest loop enclosing the statement
+        * The parent of the closest loop
+    """
+
+    class StmtInfo(object):
+        """Simple container class defining ``StmtTracker`` values."""
+
+        INFO = ['stmt', 'decl', 'loop', 'place']
+
+        def __init__(self, **kwargs):
+            for k, v in kwargs.iteritems():
+                assert(k in self.__class__.INFO)
+                setattr(self, k, v)
+
+    def __init__(self):
+        super(StmtTracker, self).__init__()
+        self.byvalue = OrderedDict()
+
+    def __setitem__(self, key, value):
+        if not isinstance(value, self.StmtInfo):
+            if not isinstance(value, tuple):
+                raise RuntimeError("StmtTracker accepts tuple or StmtInfo objects")
+            value = self.StmtInfo(**dict(zip(self.StmtInfo.INFO, value)))
```

Comment:
`zip` throws away entries from the longer list.  Do you want to assert that `len(value) == len(self.StmtInfo.INFO`?